### PR TITLE
Subject specific dates SE-1804

### PR DESCRIPTION
**Pull request by rjlynch**
### Context
Schools want to be able to limit what subjects are available on a particular date.

### Changes proposed in this pull request
* Placement dates now have a flag `supports_subjects?` Placement dates for primary schools do not support subjects, placement dates for secondary and colleges do.
Schools that are both primary and secondary are now given an option to identify a placement date as supporting subjects.

* Candidate registration wizard now includes conditional branching depending on whether a step is required for the current school they are applying for. Eg for a school with fixed dates the wizard now starts on the select placement date screen. For flexible date schools they start on the personal information page.

* Candidates applying to a school with fixed dates will now be presented with a list of dates. If subjects are associated with dates candidates will be able to choose the subject/date combination. Primary dates are separately listed on the same page.

* bookings_placement_date_id has moved nesting in the registration_session, it is now under subject_and_date_information. We can use this key to identify in flight registration_sessions and map them to a non subject specific date see `RegistrationSession#subject_and_date_information`

* fixed dates have been moved from the right hand column of the school profile page to the main column and are placed above the 'Start request' button

* the chosen subject has been added to the check your answers screen

### Guidance to review
Manual testing steps:
From the school dashboard, click manage dates and set up a school with subject specific dates. Apply to this school, accept the placement request as the school expect the subject selected in the confirm booking screen to match the subject for the date you applied to.
Switch the school to flexible dates, make sure applying to the school still functions as before.
Repeat this test for primary, secondary, and primary + secondary.

### CODE REVIEW

File: app/controllers/candidates/registrations/subject_and_date_informations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        @school = current_registration.school
+      end
+
+      def subject_and_date_params
+        params.require(:candidates_registrations_subject_and_date_information).permit(:subject_and_date_ids)
+      end
+
+      def extract_subject_and_date_params
+        bookings_placement_date_id, bookings_placement_dates_subject_id = \
+          *subject_and_date_params.dig('subject_and_date_ids').split('_')
```
Comment (jebw) [inline app/controllers/candidates/registrations/subject_and_date_informations_controller.rb]: There's some comments elsewhere in the review about restructuring how the combined id is handled. 

That will then simplify this to allow passing `subject_and_date_params` into `SubjectAndDateInformation.new`

File: app/views/schools/placement_dates/new.html.erb [hunk 2]
```diff
@@ -23,6 +23,10 @@
   <%= GovukElementsErrorsHelper.error_summary form.object, 'There is a problem', '' %>
   <%= form.date_field :date, heading: true %>
   <%= form.number_field :duration, width: 3, required: true, label_options: { class: 'govuk-heading-m', overwrite_defaults!: true } %>
+
+  <% if show_subject_support_option(@current_school) %>
+    <%= form.radio_button_fieldset :supports_subjects, choices: [false, true] %>
```
Comment (jebw) [inline app/views/schools/placement_dates/new.html.erb]: This is defaulting to 'secondary' - should it be? Shouldn't it be just forcing the user to choose one or the other via validation?

Comment (rjlynch) [reply app/views/schools/placement_dates/new.html.erb]: working on this

File: app/controllers/schools/placement_dates_controller.rb [hunk 3]
```diff
@@ -21,6 +21,12 @@ def create
       .bookings_placement_dates
       .new(new_placement_date_params)
 
+    # if the user hasn't seen the 'select a phase' option on #new, set
+    # it here based on their available phases
+    if !new_placement_date_params.has_key?(:supports_subjects)
```
[the remark this answers is not in the document]
Comment (rjlynch) [reply app/controllers/schools/placement_dates_controller.rb]: working on this

File: app/views/candidates/registrations/placement_preferences/_form.html.erb [hunk 4]
```diff
@@ -7,15 +7,7 @@
...
-      <%- if @school.availability_preference_fixed? %>
-
-        <%= f.radio_button_fieldset :bookings_placement_date_id,
-          choices: @placement_dates,
-          value_method: :id,
-          text_method: :to_s
-        -%>
-
-      <%- else -%>
+      <%- if !@school.availability_preference_fixed? %>
```
[the remark this answers is not in the document]
Comment (rjlynch) [reply app/views/candidates/registrations/placement_preferences/_form.html.erb]: working on this

File: app/presenters/candidates/school_presenter.rb [hunk 5]
```diff
@@ -75,12 +75,39 @@ def dbs_policy
...
+        .bookings_placement_dates
+        .in_date_order
+        .supporting_subjects
+        .eager_load(:placement_date_subjects, :subjects).available
+    end
+
+    def secondary_dates_grouped_by_date
+      secondary_dates
+        .group_by(&:date)
+        .each
```
[the remark this answers is not in the document]
Comment (peteryates) [reply app/presenters/candidates/school_presenter.rb]: Well spotted, thanks

File: app/controllers/schools/placement_requests/acceptance/confirm_booking_controller.rb [hunk 6]
```diff
@@ -6,7 +6,7 @@ class ConfirmBookingController < Schools::BaseController
         before_action :set_available_subjects
 
         def new
-          subject_id = Bookings::Subject.find_by(name: @placement_request.subject_first_choice).id
+          subject_id = Bookings::Subject.find_by(name: @placement_request.requested_subject).id
```
[the remark this answers is not in the document]
Comment (rjlynch) [reply app/controllers/schools/placement_requests/acceptance/confirm_booking_controller.rb]: working on this

File: app/controllers/schools/placement_requests_controller.rb [hunk 7]
```diff
@@ -22,7 +22,7 @@ def placement_requests
       current_school
         .placement_requests
         .unbooked
-        .eager_load(:candidate, :candidate_cancellation, :school_cancellation, :placement_date, :booking)
+        .eager_load(:candidate, :candidate_cancellation, :school_cancellation, :placement_date, :booking, :subject)
```
[the remark this answers is not in the document]
Comment (peteryates) [reply app/controllers/schools/placement_requests_controller.rb]: I'd be inclined to agree when there's a large number of rows being pulled in but all of these are either `belongs_to` or `has_one`. Let's leave it for the moment but I'm happy to split it up later if required

File: app/services/candidates/registrations/subject_and_date_information.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+      def bookings_subject
+        @bookings_subject ||= Bookings::Subject.find_by(id: bookings_subject_id)
+      end
+
+      def subject_and_date_ids
+        [bookings_placement_date_id, bookings_placement_dates_subject_id].compact.join('_')
+      end
+
+      def primary_placement_dates
```
[the remark this answers is not in the document]
Comment (rjlynch) [reply app/services/candidates/registrations/subject_and_date_information.rb]: working on this

File: app/services/candidates/registrations/subject_and_date_information.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+          .bookings_placement_dates
+          .in_date_order
+          .not_supporting_subjects
+          .map do |date|
+            PlacementDateOption.new(date.id, nil, date.date.to_formatted_s(:govuk), date.duration)
+          end
+      end
+
+      def secondary_placement_dates_grouped_by_date
+        school
```
[the remark this answers is not in the document]
Comment (rjlynch) [reply app/services/candidates/registrations/subject_and_date_information.rb]: working on this

**Comment by peteryates** [comment]
>     1. We have an N+1 on Schools#show - you can see the errors in the rails logs - triggered for a school with both primary and secondary subject specific dates

Just addressed this in d926fc8

>     2. input[type=radio][choices=????] this is on SubjectAndDateInformationsController#new when choosing dates - looks like rails is trying to serialize the struct somehow and just pushing out internal ids

Fixed this in https://github.com/DFE-Digital/govuk_elements_form_builder/pull/32

I'll bump the version in a mo

File: db/migrate/20191029134926_remove_bookings_placement_dates_subjects_from_bookings_placement_requests.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class RemoveBookingsPlacementDatesSubjectsFromBookingsPlacementRequests < ActiveRecord::Migration[5.2]
+  require_relative '20191015152254_add_bookings_placement_dates_subject_id_to_bookings_placement_requests'
```
[the remark this answers is not in the document]
Comment (jebw) [reply db/migrate/20191029134926_remove_bookings_placement_dates_subjects_from_bookings_placement_requests.rb]: Not sure why this had been marked resolved - there was no comment - as far as I know these migrations still need updating? Thanks

[the remark this answers is not in the document]
Comment (peteryates) [reply db/migrate/20191029134926_remove_bookings_placement_dates_subjects_from_bookings_placement_requests.rb]: Ah my bad, thought Rich had been done. I'll fix this now

**Comment by peteryates** [comment]
# 🎉

**Pull request merged**